File: week1/hw1/hw1prob2.py
# (c) 2023 Michael Fitzgerald (mpfitz@ucla.edu)
#
# Performing some number-crunching as a componnet of a comparison between Python and C..
#
# This code shows an example of the Python 'vectorize' decorator, as well as setting
# up calculation gris and saving an array to a binary file.
#

import numpy as np

max_iter = 100 # maximum number of iterations
r2_max = 1 << 16 # 2e8
log2 = np.log(2)

@np.vectorize # this allows us to pass arrays as inputs, even though the function is written for scalars
def calc_val(x0, y0):
    ii = 0 # iteration counter
    x, y = 0., 0.

    # iteration calculator
    while (x*x + y*y <= r2_max) and (ii < max_iter):
        xt = x*x - y*y + x0
        y = 2*x*y + y0
        x = xt
        ii += 1

    # if we hit the r2_max criterion, make an adjustment
    if ii < max_iter:
        log_zn = np.log(x*x + y*y) / 2.
        nu = np.log(log_zn / log2) / log2
        ii = ii + 1 - nu

    return ii

# set up calculation grid
xmin, xmax = -2.5, 1.
ymin, ymax = -1., 1.
ny, nx = 768, 1024
xx = np.linspace(xmin, xmax, nx, endpoint=True)
yy = np.linspace(ymin, ymax, ny, endpoint=True)
xx, yy = np.meshgrid(xx, yy)

# run the calculation
zz = calc_val(xx, yy) # uses vectorized version so can take arrays as inputs

# save output data to binary file
out_fn = 'hw1prob2.npy'
np.save(out_fn, zz)

File: week1/hw1/hw1prob2.c
/*
 *
 * (c) 2023 Michael Fitzgerald (mpfitz@ucla.edu)
 *
 * This performs some number-crunching as a component of a comparison between Python and C.
 *
 *
 */

#include <stdlib.h>
#include <stdio.h>
#include <math.h>

// these are pre-processing variables
#define NX 1024
#define NY 768
#define XMIN -2.5
#define XMAX 1.
#define YMIN -1.
#define YMAX 1.
#define XSTEP (XMAX-XMIN)/(NX-1)
#define YSTEP (YMAX-YMIN)/(NY-1)
#define MAX_ITER 100 // maximum number of iterations
#define R2_MAX 1<<16 // 2e8

double calc_val(double x0, double y0) {
  double ii = 0; // iteration counter; not an int because we modify it in 2nd block below
  double x = 0.;
  double y = 0.;
  double xt, log_zn, nu;
  double log2 = log(2.);

  while ((x*x + y*y <= R2_MAX) && (ii < MAX_ITER)) {
    xt = x*x - y*y + x0;
    y = 2*x*y + y0;
    x = xt;
    ii++;
  }

  if (ii < MAX_ITER) {
    log_zn = log(x*x + y*y) / 2.;
    nu = log(log_zn / log2) / log2;
    ii = ii + 1 - nu;
  }
  
  return ii;
}

int main(int argc, char* argv[]) {
  // set up grid
  double XX[NX];
  double YY[NY];
  double ZZ[NY][NX];
  for (int i=0; i<NX; i++) XX[i] = XMIN+i*XSTEP;
  for (int i=0; i<NY; i++) YY[i] = YMIN+i*YSTEP;

  // run calculation
  for (int i=0; i<NY; i++)
    for (int j=0; j<NX; j++)
      ZZ[i][j] = calc_val(XX[j], YY[i]);

  // output to binary file
  static char* fn = "hw1prob2.dat";
  FILE *f = fopen(fn, "wb");
  fwrite(ZZ, sizeof(double), NX*NY, f);
  fclose(f);

  // cleanup
  return 0;
}
